File: datasets/C++/Claude/DataStructures/easy4.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>
#include <sstream>
#include <algorithm>
using namespace std;

vector<int> getMax(vector<string>& operations) {
    stack<int> mainStack;
    stack<int> maxStack;
    vector<int> results;
    
    for (const string& operation : operations) {
        stringstream ss(operation);
        int opType;
        ss >> opType;
        
        if (opType == 1) {  // Push element
            int x;
            ss >> x;
            mainStack.push(x);
            
            // Update maxStack
            if (maxStack.empty()) {
                maxStack.push(x);
            } else {
                maxStack.push(max(x, maxStack.top()));
            }
            
        } else if (opType == 2) {  // Pop element
            if (!mainStack.empty()) {
                mainStack.pop();
                maxStack.pop();
            }
            
        } else if (opType == 3) {  // Get maximum
            if (!maxStack.empty()) {
                results.push_back(maxStack.top());
            }
        }
    }
    
    return results;
}

int main() {
    int n;
    cout << "Enter the number of operations: ";
    cin >> n;
    cin.ignore(); // Clear the newline from buffer
    
    vector<string> operations(n);
    cout << "Enter " << n << " operations:" << endl;
    cout << "Format: '1 x' for push, '2' for pop, '3' for get max" << endl;
    
    for (int i = 0; i < n; i++) {
        getline(cin, operations[i]);
    }
    
    vector<int> results = getMax(operations);
    
    cout << "\nResults for type 3 operations (get maximum):" << endl;
    for (int result : results) {
        cout << result << endl;
    }
    
    return 0;
}

File: datasets/Python/Claude/DataStructures/easy4.py
def getMax(operations):
    """
    Process stack operations and return results for type 3 queries (get maximum).
    
    Algorithm:
    - Use two stacks: main_stack for elements, max_stack for tracking maximums
    - For push: add to main_stack, add max(current_element, top_of_max_stack) to max_stack
    - For pop: remove from both stacks
    - For get_max: return top of max_stack
    """
    main_stack = []
    max_stack = []
    results = []
    
    for operation in operations:
        parts = operation.strip().split()
        op_type = int(parts[0])
        
        if op_type == 1:  # Push element
            x = int(parts[1])
            main_stack.append(x)
            
            # Update max_stack: new max is either current element or previous max
            if not max_stack:
                max_stack.append(x)
            else:
                max_stack.append(max(x, max_stack[-1]))
                
        elif op_type == 2:  # Pop element
            if main_stack:
                main_stack.pop()
                max_stack.pop()
                
        elif op_type == 3:  # Get maximum
            if max_stack:
                results.append(max_stack[-1])
    
    return results

def main():
    print("Enter the number of operations:")
    n = int(input())
    
    operations = []
    print(f"Enter {n} operations:")
    print("Format: '1 x' for push, '2' for pop, '3' for get max")
    
    for i in range(n):
        operation = input().strip()
        operations.append(operation)
    
    results = getMax(operations)
    
    print("\nResults for type 3 operations (get maximum):")
    for result in results:
        print(result)

if __name__ == "__main__":
    main()